### PR TITLE
Support reading INT96 timestamps as LongTimestampWithTimeZoneType

### PR DESCRIPTION
**Pull request by raunaqmorarka**
Hive tables migrated to iceberg through Spark require the ability to read INT96 timestamps as LongTimestampWithTimeZoneType

## Description
Fixes https://github.com/trinodb/trino/issues/11338

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
# Iceberg
* Fix failure to read hive tables migrated to iceberg through Apache Spark. ({issue}`11338`)
```

### CODE REVIEW

**Review by wendigo — APPROVED** [review]
Ship it!

File: lib/trino-parquet/src/main/java/io/trino/parquet/reader/decoders/ValueDecoders.java [hunk 1]
```diff
@@ -469,6 +469,32 @@ public void skip(int n)
...
+                    for (int i = offset; i < offset + length; i++) {
+                        long epochSeconds = decodeFixed12First(values, i);
+                        int nanosOfSecond = decodeFixed12Second(values, i);
+                        if (precision < 9) {
+                            nanosOfSecond = (int) round(nanosOfSecond, 9 - precision);
+                        }
+                        long utcMillis = epochSeconds * MILLISECONDS_PER_SECOND + floorDiv(nanosOfSecond, NANOSECONDS_PER_MILLISECOND);
+                        encodeFixed12(
+                                packDateTimeWithZone(utcMillis, UTC_KEY),
+                                floorMod(nanosOfSecond, NANOSECONDS_PER_MILLISECOND) * PICOSECONDS_PER_NANOSECOND,
```
Comment (findinpath) [inline lib/trino-parquet/src/main/java/io/trino/parquet/reader/decoders/ValueDecoders.java:491]: Why is this different from 

https://github.com/trinodb/trino/blob/eda8fdfef81e83bb81c09f3c99d76c75a70e6558/lib/trino-parquet/src/main/java/io/trino/parquet/reader/decoders/ValueDecoders.java#L429

File: lib/trino-parquet/src/test/java/io/trino/parquet/reader/TestingColumnReader.java [hunk 2]
```diff
@@ -622,6 +629,26 @@ private static Assertion<Number> assertLongTimestampWithTimeZoneNanos(int precis
...
+    {
+        TimestampWithTimeZoneType type = createTimestampWithTimeZoneType(precision);
+        return (values, block, offset, blockOffset) ->
+        {
+            long epochSeconds = values[offset].epochSeconds();
+            int nanos = values[offset].nanosOfSecond();
+            long epochNanos = epochSeconds * Timestamps.NANOSECONDS_PER_SECOND + nanos;
+            if (precision < 9) {
+                epochNanos = round(epochNanos, 9 - precision);
+            }
```
Comment (findinpath) [inline lib/trino-parquet/src/test/java/io/trino/parquet/reader/TestingColumnReader.java:642]: Is this code section tested anywhere?

have `.123456789` in the storage , but having `MICROSECONDS` or `MILLISECONDS` the precision being read?

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java [hunk 3]
```diff
@@ -8804,19 +8804,33 @@ public void testSelectFromPrestoViewReferencingHiveTableWithTimestamps()
...
+            case MILLISECONDS -> "123";
+            case MICROSECONDS -> "123456";
+            case NANOSECONDS -> "123456789";
+        };
 
-        assertUpdate("CREATE TABLE test_timestamptz_base (t timestamp) WITH (format = 'PARQUET')");
-        assertUpdate("INSERT INTO test_timestamptz_base (t) VALUES" +
-                     "(timestamp '2022-07-26 12:13')", 1);
+        assertUpdate(session, "CREATE TABLE test_timestamptz_base (t timestamp(" + timestampPrecision.getPrecision() + ")) WITH (format = 'PARQUET')");
+        assertUpdate(session, "INSERT INTO test_timestamptz_base (t) VALUES" +
```
Comment (findinpath) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java:8824]: let's write all of the fractional part values into the storage `.123` , `.123456`, `.123456789`
to verify that the right value is being read also when using a smaller precision in the session

File: lib/trino-parquet/src/test/java/io/trino/parquet/reader/TestingColumnReader.java (hunk 2 above)
Comment (raunaqmorarka) [reply lib/trino-parquet/src/test/java/io/trino/parquet/reader/TestingColumnReader.java:642]: Yes, this is called when TIMESTAMP_TZ_MICROS is tested, `io.trino.parquet.reader.TestingColumnReader#WRITE_INT96` populates nanos which result in rounding to lower precision

File: lib/trino-parquet/src/main/java/io/trino/parquet/reader/decoders/ValueDecoders.java (hunk 1 above)
Comment (raunaqmorarka) [reply lib/trino-parquet/src/main/java/io/trino/parquet/reader/decoders/ValueDecoders.java:491]: It's the same thing either way, we can skip floorDiv/floorMod and use division and modulo directly as `nanosOfSecond` is guaranteed to be positive by this point

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java (hunk 3 above)
Comment (raunaqmorarka) [reply plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java]: updated the test

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java [hunk 4]
```diff
@@ -8837,10 +8853,23 @@ public void testTimestampWithTimeZone()
...
+            case NANOSECONDS -> 123456789;
+        };
+        assertThat(query(session, "SELECT * FROM test_timestamptz"))
+                .matches("VALUES " +
+                        "(TIMESTAMP '2022-07-26 17:13:14.123 UTC')," +
+                        "(TIMESTAMP '2022-07-26 17:13:14." + microsFraction + " UTC')," +
+                        "(TIMESTAMP '2022-07-26 17:13:14." + nanosFraction + " UTC')");
 
         assertUpdate("DROP TABLE test_timestamptz");
+        assertUpdate("DROP TABLE test_timestamptz_base");
```
Comment (ebyhr) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java:8872]: We could add a similar test in Iceberg connector. e.g. 
```java
    @Test // TestIcebergV2 
    public void testInt96TimestampWithTimeZone()
    {
        assertUpdate("CREATE TABLE hive.tpch.test_timestamptz_base (t timestamp) WITH (format = 'PARQUET')");
        assertUpdate("INSERT INTO hive.tpch.test_timestamptz_base (t) VALUES" +
                "(timestamp '2022-07-26 12:13')", 1);

        // Writing TIMESTAMP WITH LOCAL TIME ZONE is not supported, so we first create Parquet object by writing unzoned
        // timestamp (which is converted to UTC using default timezone) and then creating another table that reads from the same file.
        String tableLocation = metastore.getTable("tpch", "test_timestamptz_base").orElseThrow().getStorage().getLocation();

        // TIMESTAMP WITH LOCAL TIME ZONE is not mapped to any Trino type, so we need to create the metastore entry manually
        metastore.createTable(
                new io.trino.metastore.Table(
                        "tpch",
                        "test_timestamptz",
                        Optional.of("hive"),
                        "EXTERNAL_TABLE",
                        new Storage(
                                HiveStorageFormat.PARQUET.toStorageFormat(),
                                Optional.of(tableLocation),
                                Optional.empty(),
                                false,
                                Collections.emptyMap()),
                        List.of(new Column("t", HiveType.HIVE_TIMESTAMPLOCALTZ, Optional.empty(), Map.of())),
                        List.of(),
                        Collections.emptyMap(),
                        Optional.empty(),
                        Optional.empty(),
                        OptionalLong.empty()),
                PrincipalPrivileges.fromHivePrivilegeInfos(Collections.emptySet()));

        assertThat(query("SELECT * FROM hive.tpch.test_timestamptz"))
                .matches("VALUES TIMESTAMP '2022-07-26 17:13:00.000 UTC'");

        String path = (String) computeScalar("SELECT \"$path\" FROM hive.tpch.test_timestamptz_base");
        long size = (long) computeScalar("SELECT \"$file_size\" FROM hive.tpch.test_timestamptz_base");

        // Read a Parquet file from Iceberg connector
        assertUpdate("CREATE TABLE iceberg.tpch.test_timestamptz_migrated(t TIMESTAMP(6) WITH TIME ZONE)");

        BaseTable table = loadTable("test_timestamptz_migrated");

        table.updateProperties()
                .set(DEFAULT_NAME_MAPPING, toJson(MappingUtil.create(table.schema())))
                .commit();

        table.newAppend()
                .appendFile(DataFiles.builder(table.spec())
                        .withPath(path)
                        .withFormat(PARQUET)
                        .withFileSizeInBytes(size)
                        .withRecordCount(1)
                        .build())
                .commit();

        assertThat(query("SELECT * FROM iceberg.tpch.test_timestamptz_migrated"))
                .matches("VALUES TIMESTAMP '2022-07-26 17:13:00.000000 UTC'");

        assertUpdate("DROP TABLE hive.tpch.test_timestamptz_base");
        assertUpdate("DROP TABLE hive.tpch.test_timestamptz");
        assertUpdate("DROP TABLE iceberg.tpch.test_timestamptz_migrated");
    }
```

**Review by ebyhr — APPROVED** [review]

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java (hunk 4 above)
Comment (raunaqmorarka) [reply plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java:8872]: added the above test, thanks

**Review by marcinsbd — APPROVED** [review]

**Review by wendigo — APPROVED** [review]
:shipit:

**Review by findinpath — APPROVED** [review]

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java (hunk 4 above)
Comment (mayankvadariya) [reply plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java:8872]: I guess this test should be enough and we don't need test in `TestHiveSparkCompatibility`

**Pull request merged**